Crate fails to compile due to missing file-operations binding

I wanted to try out the `hello-world` example in the repo. Unfortunately the build fails with the following output:
```
make -C /lib/modules/4.4.0-164-generic/build M=/home/robin/code/linux-kernel-module-rust/hello-world
make[1]: Entering directory '/usr/src/linux-headers-4.4.0-164-generic'
cd /home/robin/code/linux-kernel-module-rust/hello-world; env -u MAKE -u MAKEFLAGS cargo build -Z build-std=core,alloc --target=x86_64-linux-kernel
   Compiling linux-kernel-module v0.1.0 (/home/robin/code/linux-kernel-module-rust)
error[E0560]: struct `bindings::bindings::file_operations` has no field named `clone_file_range`
   --> /home/robin/code/linux-kernel-module-rust/src/chrdev.rs:201:13
    |
201 |             clone_file_range: None,
    |             ^^^^^^^^^^^^^^^^ `bindings::bindings::file_operations` does not have this field
    |
    = note: available fields are: `owner`, `llseek`, `read`, `write`, `read_iter` ... and 22 others

error[E0560]: struct `bindings::bindings::file_operations` has no field named `copy_file_range`
   --> /home/robin/code/linux-kernel-module-rust/src/chrdev.rs:203:13
    |
203 |             copy_file_range: None,
    |             ^^^^^^^^^^^^^^^ `bindings::bindings::file_operations` does not have this field
    |
    = note: available fields are: `owner`, `llseek`, `read`, `write`, `read_iter` ... and 22 others

error[E0560]: struct `bindings::bindings::file_operations` has no field named `dedupe_file_range`
   --> /home/robin/code/linux-kernel-module-rust/src/chrdev.rs:205:13
    |
205 |             dedupe_file_range: None,
    |             ^^^^^^^^^^^^^^^^^ `bindings::bindings::file_operations` does not have this field
    |
    = note: available fields are: `owner`, `llseek`, `read`, `write`, `read_iter` ... and 22 others

error[E0560]: struct `bindings::bindings::file_operations` has no field named `iterate_shared`
   --> /home/robin/code/linux-kernel-module-rust/src/chrdev.rs:215:13
    |
215 |             iterate_shared: None,
    |             ^^^^^^^^^^^^^^ `bindings::bindings::file_operations` does not have this field
    |
    = note: available fields are: `owner`, `llseek`, `read`, `write`, `read_iter` ... and 22 others

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0560`.
error: Could not compile `linux-kernel-module`.

To learn more, run the command again with --verbose.
/home/robin/code/linux-kernel-module-rust/hello-world/Kbuild:7: recipe for target '/home/robin/code/linux-kernel-module-rust/hello-world/target/x86_64-linux-kernel/debug/libhello_world.a' failed
make[2]: *** [/home/robin/code/linux-kernel-module-rust/hello-world/target/x86_64-linux-kernel/debug/libhello_world.a] Error 101
Makefile:1454: recipe for target '_module_/home/robin/code/linux-kernel-module-rust/hello-world' failed
make[1]: *** [_module_/home/robin/code/linux-kernel-module-rust/hello-world] Error 2
make[1]: Leaving directory '/usr/src/linux-headers-4.4.0-164-generic'
Makefile:4: recipe for target 'all' failed
make: *** [all] Error 2
```

Current rust toolchain:
```
nightly-x86_64-unknown-linux-gnu (directory override for '/home/robin/code/linux-kernel-module-rust')
rustc 1.39.0-nightly (eceec57f7 2019-09-18)
```

gcc version:
```
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
```

clang version:
```
clang version 3.8.0-2ubuntu4 (tags/RELEASE_380/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

kernel:
```
Linux [...] 4.4.0-164-generic #192-Ubuntu SMP Fri Sep 13 12:02:50 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```


From what I understand my kernel version doesn't have all the file operations used in the the initialization of the `file_operations` struct in `chrdev.rs:193`.  
If i delete the missing fields i get the following error:
```
error[E0063]: missing field `aio_fsync` in initializer of `bindings::bindings::file_operations`
   --> /home/robin/code/linux-kernel-module-rust/src/chrdev.rs:193:30
    |
193 |         FileOperationsVtable(bindings::file_operations {
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^ missing `aio_fsync`
```

adding `aio_fsync: None` then results in:
```
error[E0013]: constants cannot refer to statics, use a constant instead
  --> /home/robin/code/linux-kernel-module-rust/src/bindings.rs:14:31
   |
14 | pub const GFP_KERNEL: gfp_t = BINDINGS_GFP_KERNEL;
   |                               ^^^^^^^^^^^^^^^^^^^

error: use of extern static is unsafe and requires unsafe function or block (error E0133)
  --> /home/robin/code/linux-kernel-module-rust/src/bindings.rs:14:31
   |
14 | pub const GFP_KERNEL: gfp_t = BINDINGS_GFP_KERNEL;
   |                               ^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(safe_extern_statics)]` on by default
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #36247 <https://github.com/rust-lang/rust/issues/36247>
   = note: extern statics are not controlled by the Rust type system: invalid data, aliasing violations or data races will cause undefined behavior

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0013`.
error: Could not compile `linux-kernel-module`.
```

--- a/README.md
+++ b/README.md
@@ -38,10 +38,14 @@ concurrently (such as by implementing locking).
 ## System requirements
 
 We're currently only running CI on Linux 4.15 (Ubuntu Xenial) on amd64,
-although we try to keep support for newer (and perhaps older) kernels
-working. Other architectures should work but are untested - see
-[#112](https://github.com/fishinabarrel/linux-kernel-module-rust/issues/112)
-for expected status.
+although we intend to support kernels from 4.4 through the latest
+release. Please report a bug (or better yet, send in a patch!) if your
+kernel doesn't work.
+
+Other architectures aren't implemented yet, but should work as long as
+there's Rust and LLVM support - see [#112]
+(https://github.com/fishinabarrel/linux-kernel-module-rust/issues/112)
+for expected status. They'll need src/c_types.rs ported, too.
 
 You'll need to have [Rust](https://www.rust-lang.org) - in particular
 Rust nightly, as we use [some unstable
@@ -54,6 +58,9 @@ you're running kernel 5.0 or newer, [you'll need Clang
 (released September 2019), which adds support for `asm goto`.
 You may need to set the `CLANG` environment variable appropriately,
 e.g., `CLANG=clang-9`.
+
+Very recent kernels may require newer versions of Clang - try Clang 11
+if older versions don't work for you.
 
 ## Building hello-world
 

--- a/build.rs
+++ b/build.rs
@@ -73,6 +73,18 @@ fn handle_kernel_version_cfg(bindings_path: &PathBuf) {
         }
     }
     let version = version.expect("Couldn't find kernel version");
+    if version >= kernel_version_code(4, 5, 0) {
+        println!("cargo:rustc-cfg=kernel_4_5_0_or_greater")
+    }
+    if version >= kernel_version_code(4, 7, 0) {
+        println!("cargo:rustc-cfg=kernel_4_7_0_or_greater")
+    }
+    if version >= kernel_version_code(4, 9, 0) {
+        println!("cargo:rustc-cfg=kernel_4_9_0_or_greater")
+    }
+    if version >= kernel_version_code(4, 13, 0) {
+        println!("cargo:rustc-cfg=kernel_4_13_0_or_greater")
+    }
     if version >= kernel_version_code(4, 15, 0) {
         println!("cargo:rustc-cfg=kernel_4_15_0_or_greater")
     }

--- a/src/file_operations.rs
+++ b/src/file_operations.rs
@@ -144,12 +144,15 @@ impl FileOperationsVtable {
                 open: Some(open_callback::<T>),
                 release: Some(release_callback::<T>),
 
+                #[cfg(not(kernel_4_9_0_or_greater))]
+                aio_fsync: None,
                 check_flags: None,
-                #[cfg(not(kernel_4_20_0_or_greater))]
+                #[cfg(all(kernel_4_5_0_or_greater, not(kernel_4_20_0_or_greater)))]
                 clone_file_range: None,
                 compat_ioctl: None,
+                #[cfg(all(kernel_4_5_0_or_greater, not(kernel_4_20_0_or_greater)))]
                 copy_file_range: None,
-                #[cfg(not(kernel_4_20_0_or_greater))]
+                #[cfg(all(kernel_4_5_0_or_greater, not(kernel_4_20_0_or_greater)))]
                 dedupe_file_range: None,
                 fallocate: None,
                 #[cfg(kernel_4_19_0_or_greater)]
@@ -160,6 +163,7 @@ impl FileOperationsVtable {
                 fsync: None,
                 get_unmapped_area: None,
                 iterate: None,
+                #[cfg(kernel_4_7_0_or_greater)]
                 iterate_shared: None,
                 #[cfg(kernel_5_1_0_or_greater)]
                 iopoll: None,

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -13,6 +13,7 @@ mod error;
 pub mod file_operations;
 pub mod filesystem;
 pub mod printk;
+#[cfg(kernel_4_13_0_or_greater)]
 pub mod random;
 pub mod sysctl;
 mod types;

